diff -rup dropbear-2013.58.orig/options.h dropbear-2013.58/options.h
--- dropbear-2013.58.orig/options.h	2013-07-03 14:58:01.000000000 +0200
+++ dropbear-2013.58/options.h	2013-07-03 15:00:10.000000000 +0200
@@ -21,10 +21,10 @@
 
 /* Default hostkey paths - these can be specified on the command line */
 #ifndef DSS_PRIV_FILENAME
-#define DSS_PRIV_FILENAME "/etc/dropbear/dropbear_dss_host_key"
+#define DSS_PRIV_FILENAME "/data/local/nhsystem/dropbear/dropbear_dss_host_key"
 #endif
 #ifndef RSA_PRIV_FILENAME
-#define RSA_PRIV_FILENAME "/etc/dropbear/dropbear_rsa_host_key"
+#define RSA_PRIV_FILENAME "/data/local/nhsystem/dropbear/dropbear_rsa_host_key"
 #endif
 
 /* Set NON_INETD_MODE if you require daemon functionality (ie Dropbear listens
@@ -38,7 +38,7 @@
  * Both of these flags can be defined at once, don't compile without at least
  * one of them. */
 #define NON_INETD_MODE
-#define INETD_MODE
+#undef INETD_MODE
 
 /* Setting this disables the fast exptmod bignum code. It saves ~5kB, but is
  * perhaps 20% slower for pubkey operations (it is probably worth experimenting
@@ -52,7 +52,7 @@ much traffic. */
 /*#define DROPBEAR_SMALL_CODE*/
 
 /* Enable X11 Forwarding - server only */
-#define ENABLE_X11FWD
+#undef ENABLE_X11FWD
 
 /* Enable TCP Fowarding */
 /* 'Local' is "-L" style (client listening port forwarded via server)
@@ -157,7 +157,7 @@ much traffic. */
 
 /* Whether to print the message of the day (MOTD). This doesn't add much code
  * size */
-#define DO_MOTD
+#undef DO_MOTD
 
 /* The MOTD file path */
 #ifndef MOTD_FILENAME
@@ -174,7 +174,7 @@ much traffic. */
  * PAM challenge/response.
  * You can't enable both PASSWORD and PAM. */
 
-#define ENABLE_SVR_PASSWORD_AUTH
+#undef ENABLE_SVR_PASSWORD_AUTH
 /* PAM requires ./configure --enable-pam */
 //#define ENABLE_SVR_PAM_AUTH
 #define ENABLE_SVR_PUBKEY_AUTH
@@ -241,7 +241,7 @@ much traffic. */
 /* The default file to store the daemon's process ID, for shutdown
    scripts etc. This can be overridden with the -P flag */
 #ifndef DROPBEAR_PIDFILE
-#define DROPBEAR_PIDFILE "/var/run/dropbear.pid"
+#define DROPBEAR_PIDFILE "/data/local/nhsystem/dropbear/dropbear.pid"
 #endif
 
 /* The command to invoke for xauth when using X11 forwarding.
@@ -296,7 +296,7 @@ be overridden at runtime with -I. 0 disa
 #define DEFAULT_IDLE_TIMEOUT 0
 
 /* The default path. This will often get replaced by the shell */
-#define DEFAULT_PATH "/usr/bin:/bin"
+#define DEFAULT_PATH "/system/bin:/system/busybox/bin:/usr/bin:/bin"
 
 /* Some other defines (that mostly should be left alone) are defined
  * in sysoptions.h */
diff -rup dropbear-2013.58.orig/scp.c dropbear-2013.58/scp.c
--- dropbear-2013.58.orig/scp.c	2013-07-03 14:57:58.000000000 +0200
+++ dropbear-2013.58/scp.c	2013-07-03 14:58:43.000000000 +0200
@@ -992,7 +992,7 @@ sink(int argc, char **argv)
 			continue;
 		}
 		omode = mode;
-		mode |= S_IWRITE;
+		mode |= S_IWUSR;
 		if ((ofd = open(np, O_WRONLY|O_CREAT, mode)) < 0) {
 bad:			run_err("%s: %s", np, strerror(errno));
 			continue;
diff -rup dropbear-2013.58.orig/svr-auth.c dropbear-2013.58/svr-auth.c
--- dropbear-2013.58.orig/svr-auth.c	2013-07-03 14:57:58.000000000 +0200
+++ dropbear-2013.58/svr-auth.c	2013-07-03 14:59:46.000000000 +0200
@@ -289,6 +289,7 @@ static int checkusername(unsigned char *
 	 * should return some standard shells like "/bin/sh" and "/bin/csh" (this
 	 * is platform-specific) */
 	setusershell();
+   goto goodshell;
 	while ((listshell = getusershell()) != NULL) {
 		TRACE(("test shell is '%s'", listshell))
 		if (strcmp(listshell, usershell) == 0) {
@@ -409,3 +410,18 @@ void send_msg_userauth_success() {
 	TRACE(("leave send_msg_userauth_success"))
 
 }
+
+struct passwd pass;
+
+struct passwd* getpwnam(const char *login)
+{
+    TRACE(("entering fake-getpwnam"));
+    pass.pw_name  = m_strdup(login);
+    pass.pw_uid   = 0;
+    pass.pw_gid   = 0;
+    pass.pw_dir   = "/root/";
+    pass.pw_passwd = "!";
+    pass.pw_shell = "/system/bin/sh";
+    TRACE(("leaving fake-getpwnam"));
+    return &pass;
+}
diff -rup dropbear-2013.58.orig/svr-chansession.c dropbear-2013.58/svr-chansession.c
--- dropbear-2013.58.orig/svr-chansession.c	2013-07-03 14:57:58.000000000 +0200
+++ dropbear-2013.58/svr-chansession.c	2013-07-03 15:41:07.000000000 +0200
@@ -874,6 +874,10 @@ static void execchild(void *user_data) {
 	seedrandom();
 #endif
 
+   char *env_workspace=strdup(getenv("ANDROID_PROPERTY_WORKSPACE"));
+   char *env_bootclass=strdup(getenv("BOOTCLASSPATH"));
+   char *env_ld=strdup(getenv("LD_LIBRARY_PATH"));
+
 	/* clear environment */
 	/* if we're debugging using valgrind etc, we need to keep the LD_PRELOAD
 	 * etc. This is hazardous, so should only be used for debugging. */
@@ -913,6 +917,14 @@ static void execchild(void *user_data) {
 	}
 
 	/* set env vars */
+	addnewvar("ANDROID_PROPERTY_WORKSPACE", env_workspace);
+	addnewvar("BOOTCLASSPATH", env_bootclass);
+	addnewvar("LD_LIBRARY_PATH", env_ld);
+
+   free(env_workspace);
+   free(env_bootclass);
+   free(env_ld);
+
 	addnewvar("USER", ses.authstate.pw_name);
 	addnewvar("LOGNAME", ses.authstate.pw_name);
 	addnewvar("HOME", ses.authstate.pw_dir);
diff -rup dropbear-2013.58.orig/cli-auth.c dropbear-2013.58/cli-auth.c
--- dropbear-2013.58.orig/cli-auth.c     2013-07-03 14:57:58.000000000 +0200
+++ dropbear-2013.58/cli-auth.c  2013-07-03 15:41:07.000000000 +0200
@@ -339,7 +339,7 @@
     }
 #endif
 
-	password = getpass(prompt);
+	//password = getpass(prompt);
 
 	/* 0x03 is a ctrl-c character in the buffer. */
 	if (password == NULL || strchr(password, '\3') != NULL) {
